### PR TITLE
feat(DENG-3462): add activity fields to support KPI metrics to baseline_last_seen views

### PR DESCRIPTION
**Pull request by kik-kik**
# feat(DENG-3462): add activity fields to support KPI metrics to `baseline_last_seen` views

This approach aims to add fields to enable us to calculate KPI metrics based on activity. The original proposal was to have a "main" view which would union all the tables and add these field (see: [Create telemetry.active_users view](https://github.com/mozilla/bigquery-etl/pull/5247) ).

This PR takes a slightly different approach and instead introduces those additional fields in the upstream view (baseline_clients_last_seen).

The following fields are being added to support our KPI metric calculations:

- `activity_segment` - categorise a client based on their activity "levels"
- `is_dau` - sent ping and considered active on current date
- `is_wau` - sent ping and considered active in the last 7 days
- `is_mau` - sent ping and considered active in the last 28 days
- `is_daily_user` - client a baseline sent ping on current date
- `is_weekly_user` - client a baseline sent ping in the 7 days
- `is_monthly_user` - client a baseline sent ping in the last 28 days

---

Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title).
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated.
- [ ] When adding a new derived dataset, ensure that data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data can be available in the [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](https://github.com/mozilla/bigquery-etl/blob/main/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-3531)

### CODE REVIEW

**Comment by kwindau** [comment]
@kik-kik do all the fields have the columns it is referencing? distribution ID, isp_name, etc?

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql [hunk 1]
```diff
@@ -6,7 +6,35 @@ AS
...
+      THEN CONCAT(app_name, ' ', distribution_id)
+    {% endif %}
+    ELSE app_name
+  END AS analysis_app_name,
+  NOT (
+    LOWER(isp) = "browserstack"
+    {% if app_name = "fenix" %}
+    AND LOWER(distribution_id) = "mozillaonline"
+    {% endif %}
+  ) AS is_mobile, -- TODO: add logic to check the app is a mobile app
```
Comment (bochocki) [inline sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:32]: Is this logic null safe? I suspect it is not:
```sql
SELECT NOT(TRUE AND x = "mozillaonline")
  FROM (SELECT CAST(NULL AS STRING) AS x)
```
returns `null`.

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql [hunk 2]
```diff
@@ -6,7 +6,35 @@ AS
...
+  mozfun.bits28.days_since_seen(days_seen_bits) < 28 AS is_monthly_user,
+  CASE
+    WHEN LOWER(isp) = 'browserstack'
+      THEN CONCAT(app_name, ' ', isp_name)
+    {% if app_name = "fenix" %}
+    WHEN LOWER(distribution_id) = 'mozillaonline'
+      THEN CONCAT(app_name, ' ', distribution_id)
+    {% endif %}
+    ELSE app_name
+  END AS analysis_app_name,
```
Comment (bochocki) [inline sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:26]: I think it's safe to leave this as `app_name`. That's consistent with how it's reported in `active_users_aggregates`:

```sql
SELECT DISTINCT app_name
  FROM `moz-fx-data-shared-prod.firefox_ios.active_users_aggregates`
 WHERE submission_date = "2024-04-04"
```
returns `Firefox iOS, Firefox iOS BrowserStack`

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql (hunk 1 above)
Comment (bochocki) [inline sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:32]: This should always be true for Fenix, Firefox iOS, Focus Android, and Focus iOS. It should always be False for Klar iOS and Klar Android.

File: sql_generators/glean_usage/templates/baseline_clients_daily.view.sql [hunk 3]
```diff
@@ -4,6 +4,12 @@ CREATE OR REPLACE VIEW
   `{{ project_id }}.{{ daily_view }}`
 AS
 SELECT
-  *
+  *,
```
Comment (lucia-vargas-a) [inline sql_generators/glean_usage/templates/baseline_clients_daily.view.sql:7]: This change is not part of the scope of the `active_users` view, could it be in a separate PR?

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql [hunk 4]
```diff
@@ -6,7 +6,35 @@ AS
...
+  mozfun.bits28.days_since_seen(days_active_bits) < 7 AS is_wau,
+  mozfun.bits28.days_since_seen(days_active_bits) < 28 AS is_mau,
+  -- Metrics based on pings sent
+  mozfun.bits28.days_since_seen(days_seen_bits) = 0 AS is_daily_user,
+  mozfun.bits28.days_since_seen(days_seen_bits) < 7 AS is_weekly_user,
+  mozfun.bits28.days_since_seen(days_seen_bits) < 28 AS is_monthly_user,
+  CASE
+    WHEN LOWER(isp) = 'browserstack'
+      THEN CONCAT(app_name, ' ', isp_name)
+    {% if app_name = "fenix" %}
```
Comment (lucia-vargas-a) [inline sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:21]: This logic of `distribution_id` applies also to Firefox Desktop.

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql [hunk 5]
```diff
@@ -6,7 +6,35 @@ AS
...
+    {% endif %}
+    ELSE app_name
+  END AS analysis_app_name,
+  NOT (
+    LOWER(isp) = "browserstack"
+    {% if app_name = "fenix" %}
+    AND LOWER(distribution_id) = "mozillaonline"
+    {% endif %}
+  ) AS is_mobile, -- TODO: add logic to check the app is a mobile app
+  FALSE AS is_desktop, -- TODO: add logic to check the app is desktop
```
Comment (lucia-vargas-a) [inline sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:33]: This template also generates `firefox_desktop.baseline_clients_last_seen` so it needs the logic for Desktop implemented.

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql [hunk 6]
```diff
@@ -6,7 +6,35 @@ AS
...
   *
 FROM
   `{{ project_id }}.{{ last_seen_table }}`
+{% if app_name = "fenix" %}
```
Comment (lucia-vargas-a) [inline sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:37]: Is there a source for `distribution_id` for Desktop (Glean) that can be also implemented here?

**Review by lucia-vargas-a — COMMENTED** [review]
Some comments and the need to account for Desktop, because the Glean template also generates desktop's view.

**Review by lucia-vargas-a — COMMENTED** [review]
This part of the description is incorrect `This is a re-implementation of https://github.com/mozilla/bigquery-etl/pull/5396`.
I'd probably suggest "This PR moves the logic to calculate growth metrics from a single view for all apps to the clients_last_seen` views per app.".

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql (hunk 1 above)
Comment (kik-kik) [reply sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:32]: correct, this is a bug.

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql (hunk 2 above)
Comment (kik-kik) [reply sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:26]: ~We have to keep in mind that this specific field in this table might be used elsewhere so we should be at the very least careful about swapping out this value.~

Actually, apologies. I just checked and app_name does not exist yet so it's fine to just call it app_name. I can see I will also need to include string values for it.

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql (hunk 5 above)
Comment (kik-kik) [reply sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:33]: yes, sorry. I made this as a draft, it was not quite ready for a review yet. I will be pushing more changes soon.

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql [hunk 7]
```diff
@@ -6,7 +6,35 @@ AS
 SELECT
   {% for ut in usage_types %}
     `moz-fx-data-shared-prod`.udf.pos_of_trailing_set_bit(days_{{ ut }}_bits) AS days_since_{{ ut }},
-  {% endfor %}
+  {% endfor %},
+  -- Metrics based on activity
```
Comment (lucia-vargas-a) [inline sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:10]: We have the logic of activity segments in the [clients_last_seen](https://github.com/mozilla/bigquery-etl/blob/15c4835f5a0e2b7d0953f612ff0f2230df789897/sql/moz-fx-data-shared-prod/telemetry/clients_last_seen_v1/view.sql#L84-L101) view for Legacy telemetry.
Could we please add it here for Glean telemetry to avoid having it hard-coded in several places and given that these are source of truth views?
```suggestion
  -- Metrics based on activity
  CASE
      WHEN BIT_COUNT(days_active_bits)
        BETWEEN 1
        AND 6
        THEN 'infrequent_user'
      WHEN BIT_COUNT(days_active_bits)
        BETWEEN 7
        AND 13
        THEN 'casual_user'
      WHEN BIT_COUNT(days_active_bits)
        BETWEEN 14
        AND 20
        THEN 'regular_user'
      WHEN BIT_COUNT(days_active_bits) >= 21
        THEN 'core_user'
      ELSE 'other'
    END AS activity_segment,
```

**Review by lucia-vargas-a — COMMENTED** [review]
Comment on adding activity segments.

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql (hunk 1 above)
Comment (kik-kik) [reply sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql]: Made a tweak to the logic.

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql (hunk 4 above)
Comment (kik-kik) [reply sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql]: Moved to a different template which does not appear to generate firefox_desktop views.

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql (hunk 6 above)
Comment (kik-kik) [reply sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql]: Ignoring as moved to a different template which does not render for firefox_desktop as it does not have multiple channel tables.

File: sql_generators/glean_usage/templates/cross_channel.view.sql [hunk 8]
```diff
@@ -7,13 +9,59 @@ AS
...
+            BETWEEN 7 AND 13
+                THEN 'casual_user'
+        WHEN BIT_COUNT(days_active_bits)
+            BETWEEN 14 AND 20
+                THEN 'regular_user'
+        WHEN BIT_COUNT(days_active_bits) >= 21
+            THEN 'core_user'
+        ELSE 'other'
+    END AS activity_segment,
+    mozfun.bits28.days_since_seen(days_active_bits) = 0 AS is_dau,
```
Comment (lucia-vargas-a) [inline sql_generators/glean_usage/templates/cross_channel.view.sql:36]:
```suggestion
    IFNULL(mozfun.bits28.days_since_seen(days_active_bits) = 0, FALSE) AS is_dau,
```

File: sql_generators/glean_usage/templates/cross_channel.view.sql [hunk 9]
```diff
@@ -7,13 +9,59 @@ AS
...
+    mozfun.bits28.days_since_seen(days_active_bits) < 7 AS is_wau,
```
Comment (lucia-vargas-a) [inline sql_generators/glean_usage/templates/cross_channel.view.sql:37]:
```suggestion
    IFNULL(mozfun.bits28.days_since_seen(days_active_bits) < 7, FALSE) AS is_wau,
```

File: sql_generators/glean_usage/templates/cross_channel.view.sql [hunk 10]
```diff
@@ -7,13 +9,59 @@ AS
...
+    mozfun.bits28.days_since_seen(days_active_bits) < 28 AS is_mau,
```
Comment (lucia-vargas-a) [inline sql_generators/glean_usage/templates/cross_channel.view.sql:38]: Required due to the bits28 UDF behaviour. Also for the columns below.
```suggestion
    mozfun.bits28.days_since_seen(days_active_bits) < 28 AS is_mau,
    IFNULL(mozfun.bits28.days_since_seen(days_active_bits) < 28, FALSE) AS is_mau,
```

**Review by lucia-vargas-a — CHANGES_REQUESTED** [review]
bits28 udf requires the IFNULL validation

File: sql_generators/glean_usage/templates/cross_channel.view.sql [hunk 11]
```diff
@@ -1,4 +1,6 @@
 -- Generated via ./bqetl generate glean_usage
+
```
Comment (lucia-vargas-a) [inline sql_generators/glean_usage/templates/cross_channel.view.sql:2]:
```suggestion
```

**Review by lucia-vargas-a — COMMENTED** [review]
Why modifying the `cross_channel` template instead of the `baseline_clients_last_seen` template?

File: sql_generators/glean_usage/templates/cross_channel.view.sql (hunk 8 above)
Comment (kik-kik) [reply sql_generators/glean_usage/templates/cross_channel.view.sql:36]: days_active_bits should never be null?

**Comment by kik-kik** [comment]
> bits28 udf requires the IFNULL validation

ok, so this specific function could result a null?

File: sql_generators/glean_usage/templates/cross_channel.view.sql (hunk 10 above)
Comment (bochocki) [reply sql_generators/glean_usage/templates/cross_channel.view.sql]: Ah, right. I think we saw this case in earlier testing. If `days_active_bits == 0` -- which can happen when the user sends a ping but never meets the "active" criteria -- `mosfun.bits28.days_since_seen` returns `NULL`. Great catch Lucia!

Comment (kik-kik) [reply sql_generators/glean_usage/templates/cross_channel.view.sql]: Got it, thanks a lot for explaining.

**Comment by kik-kik** [comment]
> @kik-kik do all the fields have the columns it is referencing? distribution ID, isp_name, etc?

@kwindau so in this case my understanding is that because this is the baseline ping all of the apps should follow the same schema structure. To be on the safe side, for now I added a list of apps for which we should include these additional fields and in those apps the fields should exist. The SQL validation should also detect in case the view is invalid.

**Comment by sean-rose** [comment]
The `deploy-changes-to-stage` CI job is now completing successfully, but the `dry-run-sql` job is failing due to some apparent issues with this PR that will need to be fixed.

**Review by kwindau — APPROVED** [review]
Looks good to me!

**Comment by kik-kik** [comment]
Blocked by: https://github.com/mozilla/bigquery-etl/pull/5497

**Comment by scholtzan** [comment]
To get this merged we'll likely need to skip stage deploys. Some of the queries that got changed depend on certain views, and some views depend on queries that got changed.
The stage deploys first deploy all queries, then all views. So in some cases query schemas cannot be determined as the upstream views haven't been deployed yet.

Stage deploys can be skipped using the `skip-stage-deploys` CI parameter and setting it to false when triggering CI manually

**Review by lucia-vargas-a — APPROVED** [review]
Only a comment on the `products_to_include_extra_activity_fields`

File: sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql [hunk 12]
```diff
@@ -1,4 +1,19 @@
 {{ header }}
+{% set products_to_include_extra_activity_fields = [
```
[the remark this answers is not in the document]
Comment (kik-kik) [reply sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql:2]: all products we want to include, if we do not have this list along with a filter all GLEAN application would have these fields added.

[the remark this answers is not in the document]
Comment (lucia-vargas-a) [reply sql_generators/glean_usage/templates/baseline_clients_last_seen.view.sql]: Got it, products not in this list. Thanks for that clarification.

**Review by bochocki — CHANGES_REQUESTED** [review]
Blocked pending discussion.

**Comment by kik-kik** [comment]
Closing in favor of https://github.com/mozilla/bigquery-etl/pull/5594
